### ISSUE TITLE
EventGroup.ts preventing callbacks from being invoked on components in Modals

### ISSUE THREAD

**Issue by mattmazzola** [OfficeDev/office-ui-fabric-react#8264]
Seems to be a bug in [EventGroup.ts](https://github.com/OfficeDev/office-ui-fabric-react/blob/c9613c6206dd9f13c4712c93e91bcbef46d18169/packages/utilities/src/EventGroup.ts) that blocks events from within a Modal. 

Can see in the video below that buttons receive the click event but somewhere in `EventGroup.js` it is terminated. I was surprised even the native button is blocked.

https://microsoft.sharepoint.com/:v:/t/convsysresearch/ETulIpY7KP9Nl9AfDLebWxsBC545qYHqZwLAe1jvbcx3hg?e=LInPgn

It seems to be related to Modal stacking order.  Odd thing is this modal is on top.
I wonder if this means the events are being captures instead of bubbled in order for it to be intercepted.

### Environment Information 
 
"office-ui-fabric-react": "5.100.0",
Chrome: Version 72.0.3626.121 (Official Build) (64-bit)

#### Actual behavior:
 
Callbacks are not invoked.
 
#### Expected behavior:
 
Callbacks should be invoked.

#### Priorities and help requested:
 
**Are you willing to submit a PR to fix**? (No)
 
**Requested priority:** (High)
 
**Products/sites affected:**
ConversationLearner

**Comment by KevinTCoughlin** [comment]
Thanks for reporting this @mattmazzola. It would be helpful to have this isolated in a Codepen / CodeSandbox for us to be able to debug and fix quickly. Would you be able to provide one so that I can ensure the fix is correct given your scenario?

Its important to note that this is reported against `5.100.0` and so its also worth I investigate if this is still an issue in `6.x`.

**Comment by mattmazzola** [comment]
I can try to recreate. How do you create a codepen that uses Modals in 5.x?

I was starting here:
https://codepen.io/mattmazzola-the-reactor/pen/xBEwrZ?editors=1010

I am using:
`<script src="//unpkg.com/office-ui-fabric-react@5.100.0/dist/office-ui-fabric-react.min.js"></script>`
But this module doesn't export the `Modal` component.  I had mentioned this somewhere else as it seems like a first-class component that should be exported.

In code we have to do this:
![image](https://user-images.githubusercontent.com/2856501/54376864-2f2fef80-4641-11e9-9833-3e3c4def03b3.png)

And deep linking inside modules not good practice.

**Comment by KevinTCoughlin** [comment]
Hrmm, I don't see this callback issue in `6.x` -- https://codepen.io/kevintcoughlin/pen/pYZEbL?editors=1010.

Setting up my dev env to test against `5.x`...

**Comment by KevinTCoughlin** [comment]
@mattmazzola I'm unable to repro your issue with component callbacks not firing when rendered within a `Modal`. See my findings above, https://github.com/OfficeDev/office-ui-fabric-react/issues/8264#issuecomment-474131092.

If you believe this issue to be related to OUFR, can you please fork the `5.x` branch I've created to demonstrate the library bug? 

Here's the branch: https://github.com/KevinTCoughlin/office-ui-fabric-react/tree/keco/debug-modal-callbacks-v5

**Comment by mattmazzola** [comment]
Yea, I wasn't able to recreate it either. I also have separate repo which tried to similar the scenario from our app more closely but wasn't able to reproduce. The modals worked as expected then.

To be clear the issue isn't simply callbacks not being invoked when in modals.  We have many modals in our app which work fine. It's something to do with the stacking order and other components that interfere with normal behavior.  For example there was issue we had while back with help panels not being closable do to the same issue but we didn't know about `EventGroup` back then.

To help me diagnose, can you explain what `EvenGroup.ts` does and why it's needed?
Is there a way for us to disable it or our app and to let all events through?

We've had previous complications with <FocusTrapZones> etc now with the `EventGroup.ts` being another issue and would want to know why Office fabric is affecting these core browser functions? I had expected it to be a collection of themed components, but it is doing much more. I assume they're necessary for some accessibility feature but when would I require them?

I would say you don't need to try to reproduce again. We ended up working around the issue by nesting the modals. (Modal C must be declared inside Modal A and Modal B even though it's shared. We know it's poor design and not very scalable but our only work around at the moment and indistinguishable from user point of view.

If I get time I will try to update my sample to try to reproduce again.

**Comment by KevinTCoughlin** [comment]
> To help me diagnose, can you explain what `EventGroup.ts` does and why it's needed?

`EventGroup` is used for a number of scenarios in `office-ui-fabric-react`, some examples are:

1. Auto disposal of bound callbacks to prevent memory leaks via `_events`, particularly in [`BaseComponent`](https://github.com/OfficeDev/office-ui-fabric-react/blob/bd191a3964734c6619fa49fcdb1d6127e2834522/packages/utilities/src/BaseComponent.ts#L144).
1. Normalize event behaviors, particularly w.r.t IE-compatibility.
1. Custom eventing abstraction for component communications

See all usages at https://github.com/OfficeDev/office-ui-fabric-react/search?l=TypeScript&q=EventGroup.

> Is there a way for us to disable it or our app and to let all events through?

I'm unaware of a global way to disable `EventGroup` aside from overriding it with an empty module or somesuch which I wouldn't advocate.

---

Ultimately I defer to @dzearing when discussing the usefulness of `EventGroup` within the context of OUFR.

**Comment by dzearing** [comment]
`EventGroup` is simply an abstraction on `addEventListener`. Your issue has nothing to do with `EventGroup`, and more to what is receiving the callback and then calling `stopPropagation`, if I understand correctly.

But all that said, I'm really having a hard time understanding what the repro is.

Fabric 6, I put a button in a modal, executes onClick no problem.
https://codepen.io/dzearing/pen/MxBmMe?editors=1010

**Comment by dzearing** [comment]
I just downgraded NPM to build Fabric 5 and this does not repro in Fabric 5.

Something is calling preventDefault or stopPropagation, likely on a captured click event. You can debug this by doing something like this in the console:

```ts
Event.prototype.stopPropagation = () => console.trace();
```

Should spit out a callstack whenever the event is stopped.

**Comment by KevinTCoughlin** [comment]
Closing as per:

>I would say you don't need to try to reproduce again.

Feel free to ping if you wish to re-open this @mattmazzola.

**Comment by mattmazzola** [comment]
I only intended to save you spending time reproducing work, not to close. It's still an issue that people should see as unresolved.

I have created a video using David's suggesting above and you can see the `stopPropagation` coming from the `FocusTrapZone`.

Video:
https://microsoft.sharepoint.com/:v:/t/convsysresearch/EV3GnuTstDdIkeb5HiN8YGsBTC41nfUYaVZhpH6VHcFVGA?e=SLP1fg

Please re-open this issue. I guess it could be better named as `FocusTrapZone` preventing callbacks than `EventGroup` but it seems they are related.

If I recall we've have/had other issues with `FocusTrapZone` related to changing the focus of inputs from the one the user chose.  If I wanted to test with these turns off I assume I would need to do more than simple make it an empty module as other things on Office Fabric are depending on these functions.
What's the best approach?
This is now affecting use regardless if the modal affected is nested (duplicated) or shared which was our only way around this before.

**Comment by mattmazzola** [comment]
Can you clarify intended behavior of `FocusTrapZone`?

What is it surrounding?
When should it be added to the page?

There was talk in Teams about the issue occurring when multiple zones are "fighting" for control, I assume one stops the events for the other and ends to breaking everything.

In the particular case I have there are 3 and I don't know why. I understand there should really only event be 1?
![image](https://user-images.githubusercontent.com/2856501/54936370-940c0500-4edf-11e9-8336-aa5d85a65959.png)

Hover over each of Focus Zones to highlight area. See small one in middle doesn't correspond to either modal.
![http://g.recordit.co/PHjbGKYQrC.gif](http://g.recordit.co/PHjbGKYQrC.gif)

**Comment by dzearing** [comment]
The `FocusTrapZone` is used to keep the focus from going outside of a bounding box. It's primary usage is for modal scenarios where hitting TAB should cycle within a dialog, and not suddenly jump behind the semi-opaque background in your screenshot.

That's why Dialog and Panel typically have this. The scenario you're running into is a layered scenario, where you have a dialog on top of a dialog on top of a dialog. I remember in Fabric 5 (which you are using) we had some issues with stacked FocusTrapZones.

@JasonGore just fixed numerous bugs within FocusTrapZone in Fabric 6, so I'd love to make sure we're good with layered scenarios like this in the latest version.

Is there anything preventing you from upgrading to a more recent release of Fabric?

**Comment by JasonGore** [comment]
It is possible with the right set of props to have multiple FocusTrapZones layered. For example, you could set `forceFocusInsideTrap` to `false` and/or `isClickableOutsideFocusTrap` to `true`. If these props are left to default, however, there really can and should only be one active FocusTrapZone on the page at a time. [Our FocusTrapZone component page](https://developer.microsoft.com/en-us/fabric#/components/focustrapzone) shows a few examples with varying props values that do and don't allow interaction outside of FocusTrapZones depending on their configuration.

The specific example shown here concerns me a bit. Since `forceFocusInsizeTrap` defaults to `true`, this seems like it's an invalid configuration where 3 FocusTrapZones would be fighting for focus on both focus change and click events.
> ![image](https://user-images.githubusercontent.com/2856501/54936370-940c0500-4edf-11e9-8336-aa5d85a65959.png)

**Comment by JasonGore** [comment]
If on the other hand FocusTrapZones are nested (i.e. not siblings), those scenarios should be working properly. We also have examples of those on our component page.

If you have any further question, it may be best to look at specific implementation (either via existing code or CodePen) to help clarify.

**Comment by JasonGore** [comment]
I've deleted a couple of comments that I could see greatly increasing confusion. Sorry if you've received notifications for them.

I think the best thing to do here is work from examples of what you're doing to help us understand if FocusTrapZone is adequate for your scenario.

The main thing we don't know here is whether the FocusTrapZones are nested or siblings. 

FocusTrapZone should work appropriately in nested scenarios, which [our component page](https://developer.microsoft.com/en-us/fabric#/components/focustrapzone) shows an example of.

If, however, if the FocusTrapZones are siblings, the behavior may or may not be predictable. It would depend on their props values.

The only way to really figure out which scenario is applicable here is to get a concrete scenario. Can you share your implementation or perhaps put it in a [CodePen](https://aka.ms/fabricpen)?

**Comment by mattmazzola** [comment]
**Related to "Upgrading to 6.x to inherit fixes for `FocusTrapZone`":**
I think this is possibility.  If it fixes the modal issues we will definately move it up priority.  It's just harder to overlay my changes so it contains 6.x upgrade and the modal issue in same change to test to see if it fixes anything.

1. Types issue on `children` which has been fixed. https://github.com/OfficeDev/office-ui-fabric-react/issues/8177
2. Modal / DOM change issue is: https://github.com/OfficeDev/office-ui-fabric-react/issues/8178
Although I have changed our CSS to try to accomodate and I think it will work.
3. Seems to be type issue with fonts, but can fix by using number directly:
Seems to be using string "100" instead of number 100 which is accepted by React types. Not sure which is more correct.
![image](https://user-images.githubusercontent.com/2856501/54940423-ed783200-4ee7-11e9-9615-d7fb9efd84bd.png)
![image](https://user-images.githubusercontent.com/2856501/54940431-f10bb900-4ee7-11e9-9d1c-d6472b53aad4.png)
![image](https://user-images.githubusercontent.com/2856501/54940470-041e8900-4ee8-11e9-8ae6-2e00ef58598a.png)

4. Other cosmetic things I've noticed:
![image](https://user-images.githubusercontent.com/2856501/54940091-44c9d280-4ee7-11e9-9177-eca9aa6ca5be.png)
I think this can also be overcome with changes to our CSS.


Going back to `FocusTrapZone` issues:

> Can you share your implementation
Yes. Our front-end is open source. I pushed the branch I was working on here so you can see:
https://github.com/Microsoft/ConversationLearner-UI/tree/mattm/log-dialog-conflicts

> or perhaps put it in a CodePen?

I don't think 5.x allows modals in code pens. The component isn't exported. Been issue I complained about a while back. It seems fixed in 6.x.  
https://github.com/OfficeDev/office-ui-fabric-react/issues/8264#issuecomment-472970568

I have tried to recreate the scenario in separate repo with similar multiple modals and issue doesn't occur. There is something different about our app but I don't know what it is yet.

**Comment by mattmazzola** [comment]
> The main thing we don't know here is whether the FocusTrapZones are nested or siblings.

How can I tell? React Dev tools seems to flatten the list, although I think they are nested in the DOM.

> FocusTrapZone should work appropriately in nested scenarios

This is would explain why duplicating/nesting the modals fixed part of the issue, but I don't think it's a good practice/requirement. The modal is supposed to be shared / pop up under many conditions and it would require copying it to every other modal that it can layer on top of. This isn't intuitive or scalable.

> It would depend on their props values.

My understanding was `FocusTrapZones` were side affects of using other components. We never get to set props of them directly.  What affects them?

**Comment by mattmazzola** [comment]
Following up with discussions elsewhere so other people can see.

- Confirmed using 6.x version does fix the issue.
- Seems to be fixed in this release: https://github.com/OfficeDev/office-ui-fabric-react/releases/tag/office-ui-fabric-react_v6.43.0
- From this PR: https://github.com/OfficeDev/office-ui-fabric-react/pull/5351

The fix/bug was related to how the hierarchy of modals affected focus trap zones behavior so we started investigating more.  There is an issue if the modals that are open are not matching the hierarchy of them in the DOM. By correcting this I was able to make it work even on 5.x (5.100).

I will try to explain the issue although there might be technical inaccuracies
**If you are using a release without the fix (Pre 6.43):** DOM order must match visibility Order

Issue:
```
|
+--- Modal A (open second)
|
+--- Modal B (open first)
  |
  +--- Modal C (open third)
```
Because Modal A was opened before Modal C and it was higher in the hierarchy it was assumed to have higher priority even though it was not visible since Modal C was on top of it.  This meant Modal A's FocusTrapZone would be preventing interaction from the child Modal C.

In our case the fix was simply to not leave Modal A open if it can be closed without interrupting the UI removing information/context the user needs. 

If the modal was needed, the hierarchy would have to match like this:

```
|
+--- Modal A
|
+--- Modal B (open first)
  |
  +--- Modal A (open second)
    |
    +--- Modal C (open third)
```

**If you are on a release containing this fix (Post 6.34) you should not have to worry about this hierarchy issue.** 
The trap zones automatically know which was opened last which makes it the top most or visible modal and it will naturally have the correct focus / event blocking behavior.

To recap solutions:
1. Fix hierarchy
2. Upgrade to 6.x

**Comment by dzearing** [comment]
Thank you for the followup @mattmazzola!
